Makefile

```diff
@@ -135,10 +139,122 @@ proto-gen: install-proto-bin install-license-bin
 
+.PHONY: install-m3x-repo
+install-m3x-repo: install-glide install-generics-bin
+	# Check if repository exists, if not get it
+	test -d $(m3x_package_path) && test -d $(m3x_package_path)/vendor || ( \
+		go get -u $(m3x_package)                                          && \
+		cd $(m3x_package_path)                                            && \
+		glide install                                                        \
+	)
+	# If does exist but not at min version then update it
+	(cd $(m3x_package_path) && git cat-file -t $(m3x_package_min_ver) > /dev/null) || go get -u $(m3x_package)
+
+# Map generation rule for client/receivedBlocksMap
+.PHONY: map-client-received-blocks-gen
+map-client-received-blocks-gen: install-m3x-repo
+	cd $(m3x_package_path) && make hashmap-gen \
+		pkg=client \
+		key_type=idAndBlockStart \
+		value_type=receivedBlocks \
+		out_dir=$(m3db_package_path)/client \
+		rename_type_prefix=receivedBlocks
+	# Rename generated map file
+	mv -f $(m3db_package_path)/client/map_gen.go $(m3db_package_path)/client/received_blocks_map_gen.go
+
+# Map generation rule for storage/block/retrieverMap
+.PHONY: map-storage-block-retriever-gen
+map-storage-block-retriever-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=block \
+		value_type=DatabaseBlockRetriever \
+		out_dir=$(m3db_package_path)/storage/block \
+		rename_type_prefix=retriever \
+		rename_constructor=newRetrieverMap \
+		rename_constructor_options=retrieverMapOptions
+	# Rename both generated map and constructor files
+	mv -f $(m3db_package_path)/storage/block/map_gen.go $(m3db_package_path)/storage/block/retriever_map_gen.go
+	mv -f $(m3db_package_path)/storage/block/new_map_gen.go $(m3db_package_path)/storage/block/retriever_new_map_gen.go
+
+# Map generation rule for storage/bootstrap/result/Map
+.PHONY: map-storage-bootstrap-result-gen
+map-storage-bootstrap-result-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=result \
+		value_type=DatabaseSeriesBlocks \
+		out_dir=$(m3db_package_path)/storage/bootstrap/result
+
+# Map generation rule for storage package maps (to avoid double build over each other
+# when generating map source files in parallel, run these sequentially)
+.PHONY: map-storage-gen
+map-storage-gen:
+	make map-storage-database-namespaces-gen
+	make map-storage-shard-gen
+
+# Map generation rule for storage/databaseNamespacesMap
+.PHONY: map-storage-database-namespaces-gen
+map-storage-database-namespaces-gen: install-m3x-repo
+	cd $(m3x_package_path) && make idhashmap-gen \
+		pkg=storage \
+		value_type=databaseNamespace \
+		out_dir=$(m3db_package_path)/storage \
+		rename_type_prefix=databaseNamespaces \
+		rename_constructor=newDatabaseNamespacesMap \
+		rename_constructor_options=databaseNamespacesMapOptions
+	# Rename both generated map and constructor files
+	mv -f $(m3db_package_path)/storage/map_gen.go $(m3db_package_path)/storage/namespace_map_gen.go
```

Comment:
more a question: this target `map-storage-database-namespaces-gen` and `map-storage-shard-gen` both generate temporary files with the same name. Does that cause issues when you're running targets in parallel?

Reply:
Cool, makes sense